### DISCUSSION TITLE
[debug_handler] is causing compilation error for correct handler

### DISCUSSION THREAD

**Discussion by eipi1** [tokio-rs/axum#2091] [Q&A] [answered]
- [x] I have looked for existing issues (including closed) about this

### Version
```
├── axum v0.6.18
│   ├── axum-core v0.3.4
├── axum-macros v0.3.7 (proc-macro)
```

### Platform
`Linux x1 5.15.0-76-generic #83-Ubuntu SMP Thu Jun 15 19:16:32 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`

### Description
I have a simple axum router -
```
let shared_state = Arc::new(Client::new());

let _: Router = Router::new().route(
    "/agg",
    get({
        let shared_state = shared_state.clone();
        move || root_agg(shared_state, Uri::from_static("/"))
    }),
);
async fn root_agg(client: Arc<Client<HttpConnector>>, uri: Uri) -> Bytes {
    let res = client.get(uri.clone()).await.unwrap();
    // let buf = hyper::body::aggregate(res).await.unwrap();
    let buf = hyper::body::to_bytes(res).await.unwrap();
    buf
}

```
It compiles without any errors.
But if I add `#[axum_macros::debug_handler]` to the handler method, it causes compilation failure.

```
#[axum_macros::debug_handler]
async fn root_agg(client: Arc<Client<HttpConnector>>, uri: Uri) -> Bytes {
    let res = client.get(uri.clone()).await.unwrap();
    // let buf = hyper::body::aggregate(res).await.unwrap();
    let buf = hyper::body::to_bytes(res).await.unwrap();
    buf
}
```
Compiler output -
```
error[E0277]: the trait bound `Arc<Client<HttpConnector>>: FromRequestParts<()>` is not satisfied
  --> src/main.rs:51:27
   |
51 | async fn root_agg(client: Arc<Client<HttpConnector>>, uri: Uri) -> Bytes {
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `FromRequestParts<()>` is not implemented for `Arc<Client<HttpConnector>>`
   |
   = note: Function argument is not a valid axum extractor. 
           See `https://docs.rs/axum/latest/axum/extract/index.html` for details
   = help: the following other types implement trait `FromRequestParts<S>`:
             <() as FromRequestParts<S>>
             <(T1, T2) as FromRequestParts<S>>
             <(T1, T2, T3) as FromRequestParts<S>>
             <(T1, T2, T3, T4) as FromRequestParts<S>>
             <(T1, T2, T3, T4, T5) as FromRequestParts<S>>
             <(T1, T2, T3, T4, T5, T6) as FromRequestParts<S>>
             <(T1, T2, T3, T4, T5, T6, T7) as FromRequestParts<S>>
             <(T1, T2, T3, T4, T5, T6, T7, T8) as FromRequestParts<S>>
           and 25 others
   = help: see issue #48214
   = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable

For more information about this error, try `rustc --explain E0277`.
warning: `playground` (bin "playground") generated 1 warning
error: could not compile `playground` (bin "playground") due to previous error; 1 warning emitted
```
[Link to sample code](https://www.rustexplorer.com/b#%2F*%0A%5Bdependencies%5D%0Atokio%20%3D%20%7B%20version%20%3D%20%22*%22%2C%20features%20%3D%20%5B%22full%22%5D%20%7D%0A%0A%5Bdependencies.axum%5D%0Aversion%20%3D%20%220.6%22%0A%0A%5Bdependencies.bytes%5D%0Aversion%20%3D%20%221.4%22%0A%0A%5Bdependencies.hyper%5D%0Aversion%20%3D%20%220.14%22%0Afeatures%20%3D%20%5B%22full%22%5D%0A%0A%5Bdependencies.serde%5D%0Aversion%20%3D%20%221.0%22%0Afeatures%20%3D%20%5B%22derive%22%5D%0A%0A%5Bdependencies.axum-macros%5D%0Aversion%20%3D%20%220.3%22%0A%0A*%2F%0A%0Ause%20axum%3A%3A%7Brouting%3A%3Aget%2C%20Router%7D%3B%0Ause%20bytes%3A%3ABuf%3B%0Ause%20bytes%3A%3ABytes%3B%0Ause%20hyper%3A%3Aclient%3A%3AHttpConnector%3B%0Ause%20hyper%3A%3A%7BClient%2C%20Uri%7D%3B%0Ause%20std%3A%3Async%3A%3AArc%3B%0A%0A%23%5Btokio%3A%3Amain%5D%0Aasync%20fn%20main()%20%7B%0A%20%20%20%20let%20shared_state%20%3D%20Arc%3A%3Anew(Client%3A%3Anew())%3B%0A%0A%20%20%20%20let%20_%3A%20Router%20%3D%20Router%3A%3Anew().route(%0A%20%20%20%20%20%20%20%20%22%2Fagg%22%2C%0A%20%20%20%20%20%20%20%20get(%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20let%20shared_state%20%3D%20shared_state.clone()%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20move%20%7C%7C%20root_agg(shared_state%2C%20Uri%3A%3Afrom_static(%22%2F%22))%0A%20%20%20%20%20%20%20%20%7D)%2C%0A%20%20%20%20)%3B%0A%0A%20%20%20%20%2F%2F%20%2F%2F%20run%20it%20with%20hyper%20on%20localhost%3A3000%0A%20%20%20%20%2F%2F%20axum%3A%3AServer%3A%3Abind(%26%220.0.0.0%3A3000%22.parse().unwrap())%0A%20%20%20%20%2F%2F%20%20%20%20%20.serve(app.into_make_service())%0A%20%20%20%20%2F%2F%20%20%20%20%20.await%0A%20%20%20%20%2F%2F%20%20%20%20%20.unwrap()%3B%0A%7D%0A%0A%23%5Baxum_macros%3A%3Adebug_handler%5D%0Aasync%20fn%20root_agg(client%3A%20Arc%3CClient%3CHttpConnector%3E%3E%2C%20uri%3A%20Uri)%20-%3E%20Bytes%20%7B%0A%20%20%20%20let%20res%20%3D%20client.get(uri.clone()).await.unwrap()%3B%0A%20%20%20%20%2F%2F%20let%20buf%20%3D%20hyper%3A%3Abody%3A%3Aaggregate(res).await.unwrap()%3B%0A%20%20%20%20let%20buf%20%3D%20hyper%3A%3Abody%3A%3Ato_bytes(res).await.unwrap()%3B%0A%20%20%20%20buf%0A%7D%0A)

**Comment by jplatte** [comment, marked as answer]
`debug_handler` can only be used for actual handler funtions. In your case, the handler is the `move || ...` closure, *not* `root_agg`. You would also get an error trying to pass `root_agg` directly to `get`.

I guess `debug_handler` could be told to ignore specific function arguments, but I'm not sure whether it's worth it since it's not a very common use case.
